### ISSUE TITLE
How does QSYM synchronize with AFL?

### ISSUE THREAD

**Issue by natalie13m** [sslab-gatech/qsym#25]
I found some information in AFL's docs parallel_fuzzing.txt that you can synchronize new inputs with parallel fuzzing just by adding a directory (let's say qsym) in output directory of fuzzing. You can put new inputs with sequential name id:nnnnnn in output/qsym/queue. I tried with pure AFL but looks like nothing in this directory has been synchronized. Can I ask how QSYM did for synchronization? I can see the maintenance of bitmap in qsym but I think that's just for selecting and narrowing new inputs.

**Comment by jakkdu** [comment]
Sorry I cannot understand what you mean `pure AFL`.

QSYM uses the mechanism in AFL that you explained for synchronization.
But even if QSYM put some inputs in the directory, the inputs might be not interesting to AFL
and AFL will not synchronize the inputs.

Moreover, AFL could be blocked due to its own fuzzing, then it will not synchronize the input.
You can check afl/.synced/qsym file.
It saves the largest id in synchronized inputs as raw integer (little endian) format. 
Hope this help you!

**Comment by natalie13m** [comment]
Well by pure AFL I mean I just run the original AFL in parallel, with one master and one slave. Then I create one extra directory in output directory, so there are three directories in output: master, slave, mydir. Then I add new inputs manually to output/mydir/queue. I was hoping that AFL could synchronize all the new inputs in mydir, however, it didn't. I think maybe this is because of what you said about AFL blocking? I don't really understand why AFL will be blocked due to its own fuzzing process. Could you share more details about this? BTW you are really kind to answer this question, thanks in advance.

**Comment by jakkdu** [comment]
You're welcome. 

If input size is too large and if you use deterministic mode, AFL can be blocked to fuzz that input. 
(It tries to change every byte in the input).

Your way to put new input is actually correct.
Could you check the sync file? 
It should be master/.synced/mydir.

The file will contain the largest id for input. 
If the number in the sync file is equal or larger than your input id,
it means that AFL tried to synchronize your input
and reject it since the code coverage from the input is already covered by existing ones.

**Comment by natalie13m** [comment]
Yes, you are right about the synchronization. I checked with master/.synced/mydir and it seems like it's working fine. I guess the new inputs that I inserted didn't really trigger new coverage. Thank you very much for your help on this!
